### PR TITLE
Side navigation

### PR DESCRIPTION
**Pull request by matthinz**
Adds section-based navigation to the side of pages:

image

On the CMS side, this adds a new set of "navigation" options to the Page collection:

image

Closes https://trello.com/c/3fNkWMq1/223-side-navigation

### CODE REVIEW

**Review by rahearn — CHANGES_REQUESTED** [review]
No major issues, but I think the code can be clarified and simplified in a couple of small cases.

Maybe a conversation for a new story, but I feel like I often see the sidenav used for anchor tags internal to the current page. Is that supported here out of the box? (I think I can see a way to extend this for that use case if not)

File: app/models/menu.rb [hunk 1]
```diff
@@ -1,22 +1,52 @@
...
+    end
+
     def include_children?
       !!@include_children
     end
 
+    def is_for_page?(page)
+      return false if page.nil?
+      page_path = Page.normalize_path(page.filename.to_s)
+      Page.normalize_path(filename.to_s) == page_path
```
[the remark this answers is not in the document]
Comment (matthinz) [reply app/models/menu.rb]: Yeah I like this suggestion

File: spec/models/menu_item_spec.rb [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+      expect(item.text).to eql("Custom nav title")
+    end
+  end
+
+  describe "#is_for_page?" do
+    let(:parent_page) {
+      parent_page = double("Parent")
+      allow(parent_page).to receive_messages(title: "Parent page", filename: "/parent")
+      parent_page
+    }
```
[the remark this answers is not in the document]
Comment (matthinz) [reply spec/models/menu_item_spec.rb]: Ah! much nicer.

**Comment by matthinz** [comment]
No support for anchoring down to specific locations on a page, but this could be extended if needed (I think we have an additional level of hierarchy we can use in the USWDS sidenav if we need to)

**Review by rahearn — CHANGES_REQUESTED** [review]
One issue, that I may have just missed on the last review.

File: app/controllers/pages_controller.rb [hunk 3]
```diff
@@ -12,19 +12,25 @@ def home
...
+
+      @show_sidebar = @page.has_sidebar? || @side_nav_items.length > 0
+
       render formats: :html
     else
       store_location_for(:user, "/#{params[:path]}")
       redirect_to root_path
     end
-  rescue Page::NotFound
-    raise ActionController::RoutingError.new("Not Found")
```
[the remark this answers is not in the document]
Comment (matthinz) [reply app/controllers/pages_controller.rb:27]: Whoops! Good catch, I deleted those when I was doing some testing and accidentally committed.

File: spec/requests/pages_spec.rb [hunk 4]
```diff
@@ -46,6 +46,14 @@
     end
   end
 
+  describe "GET missing page" do
+    it "raises a routing error" do
```
Comment (rahearn) [inline spec/requests/pages_spec.rb:50]: [praise] 🎉

**Review by rahearn — APPROVED** [review]
🎉 

We'll review this at Wednesday's tech meeting.

**Comment by alvessatnih** [comment]
Looks good to me, too.  Thank you for looping us in, so we can see how this will work moving forward.

**Comment by rahearn** [comment]
- [ ] The business and/or user value of a story has been met
- [ ] CI pipeline green
- [ ] Code is meaningfully tested
- [ ] UI meets WCAG 2.1 - Levels A and AA
- [ ] Code has deployed to a staging environment
- [ ] [Architectural Decision Records](https://adr.github.io/) are written for major infrastructure decisions with [the Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md).
- [ ] Boundary and Data Flow Diagrams are updated when a change invalidates the current version.
- [ ] Logical Data Model Diagram is updated on DB schema changes.
- [ ] Public API methods are documented when added or changed.
- [ ] OSCAL templates are completed when security controls are implemented or modified.

**Review by ericaWilli — APPROVED** [review]

**Pull request merged**